### PR TITLE
Add Linux Networking counters

### PR DESCRIPTION
**Pull request by fearful-symmetry**
See  https://github.com/elastic/beats/issues/12210

Started this before I left for a 4-day weekend so it might be a tad all over, hopefully not.

I fear this is a case of me being a bit too 'clever' since I came up with a way to have one method that populates all of the structs. It would be possible to combine `fillStruct` and `readAndParseNetFile` , but then we would need `FieldbyName` as opposed to `Field` to get the struct, which means each call would iterate over the struct `NumField()` times as `FieldByName` just iterates until it gets a given name. I would still like to find a better way to do this.

Other than that, this is fairly straight forward. The mapping of struct field -> map is because the network counter types are known, but there's basically no documentation on the counters themselves, they're just random kernel fields. I think it would be better to have this library be more neutral, and any editorializing/filtering can be done in metricbeat.

Ah! Still need to update the readme.

### CODE REVIEW

File: providers/linux/procnet.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+		}
+		counters[keyArr[iter]] = parsed
+	}
+	return counters, nil
+}
+
+// readAndParseNetFile parses an entire file, and returns a 2D map, representing how files are sorted by protocol
+func readAndParseNetFile(body string) (map[string]map[string]int64, error) {
+	fileMetrics := make(map[string]map[string]int64)
+	bodySplit := strings.Split(strings.TrimSpace(body), "\n")
```
Comment (jsoriano) [inline providers/linux/procnet.go]: I like to use Scanners to read files (or any reader) line by line, but in these small files there is probably not much difference.

File: providers/linux/host_linux_test.go [hunk 2]
```diff
@@ -71,3 +71,20 @@ func TestHostVMStat(t *testing.T) {
...
+func TestHostNetworkCounters(t *testing.T) {
+	host, err := newLinuxSystem("testdata/fedora30").Host()
+	if err != nil {
+		t.Fatal(err)
+	}
+	s, err := host.(types.NetworkCounters).NetworkCounters()
+	if err != nil {
+		t.Fatal(err)
+	}
+
```
Comment (jsoriano) [inline providers/linux/host_linux_test.go]: Could we add some assertion on the resulting counters?

File: providers/linux/procnet_test.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+Udp: 16755 33 0 16829 0 0 0 0
+UdpLite: InDatagrams NoPorts InErrors OutDatagrams RcvbufErrors SndbufErrors InCsumErrors IgnoredMulti
+UdpLite: 0 0 0 0 0 0 0 0`
+	mapStr, err := readAndParseNetFile(file)
+	if err != nil {
+		t.Fatalf("error: %s", err)
+	}
+	testOut := types.SNMP{}
+	fillStruct(&testOut, mapStr)
+	fmt.Printf("Struct: %#v\n", testOut)
```
Comment (jsoriano) [inline providers/linux/procnet_test.go]: Same here, could we add some assertion on the resulting object?

File: providers/linux/procnet.go (hunk 1 above)
Comment (fearful-symmetry) [reply providers/linux/procnet.go]: Yah, I figured by reading the whole thing in and then checking length, I can make sure nothing has gone wrong, and it's not a huge file like you said.

File: providers/linux/host_linux.go [hunk 4]
```diff
@@ -81,6 +81,30 @@ func (h *host) VMStat() (*types.VMStatInfo, error) {
 	return parseVMStat(content)
 }
 
+// NetworkCounters reports data from /proc/net on linux
+func (h *host) NetworkCounters() (*types.NetworkCountersInfo, error) {
+
```
Comment (andrewkroh) [inline providers/linux/host_linux.go]: Can you please remove the newline from the beginning of the function.

File: providers/linux/testdata/fedora30/proc/stat [hunk 5]
```diff
@@ -0,0 +1,12 @@
+cpu  31608779 223223 15924288 1134746753 223880 4348430 1995355 266903 0 0
```
Comment (andrewkroh) [inline providers/linux/testdata/fedora30/proc/stat]: I don't see where this file is being used in the PR. If it's not used then let's remove it.

File: providers/linux/host_linux.go [hunk 6]
```diff
@@ -81,6 +81,30 @@ func (h *host) VMStat() (*types.VMStatInfo, error) {
...
+	snmpRaw, err := ioutil.ReadFile(h.procFS.path("net/snmp"))
+	if err != nil {
+		return nil, err
+	}
+	snmp, err := getNetSnmpStats(snmpRaw)
+	if err != nil {
+		return &types.NetworkCountersInfo{}, err
```
Comment (andrewkroh) [inline providers/linux/host_linux.go]: I think this should return a nil value on error. It seem inconsistent to have it both ways.

File: providers/linux/procnet.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	}
+}
+
+// parseEntry parses two lines from the net files, the first line being keys, the second being values
+func parseEntry(line1, line2 string) (map[string]int64, error) {
+	keyArr := strings.Split(strings.TrimSpace(line1), " ")
+	valueArr := strings.Split(strings.TrimSpace(line2), " ")
+
+	if len(keyArr) != len(valueArr) {
+		return nil, errors.New("Key and value lines are mismatched")
```
Comment (andrewkroh) [inline providers/linux/procnet.go]: Error messages should not begin with capitals. https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Comment (andrewkroh) [reply providers/linux/procnet.go]: There are a few of these in other places.

File: providers/linux/procnet.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	counters := make(map[string]int64)
```
Comment (andrewkroh) [inline providers/linux/procnet.go]: It looks like you know the length at this point so you can include it in the allocation (`make(map[string]int64, len(valueArr))`).

File: providers/linux/procnet.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error parsing string to int in line: %#v", valueArr)
+		}
+		counters[keyArr[iter]] = parsed
+	}
+	return counters, nil
+}
+
+// readAndParseNetFile parses an entire file, and returns a 2D map, representing how files are sorted by protocol
+func readAndParseNetFile(body string) (map[string]map[string]int64, error) {
```
Comment (andrewkroh) [inline providers/linux/procnet.go]: The function name is a little misleading IMO. It doesn't actually read any file (like disk IO wise). It just parses the netstat data.

File: providers/linux/testdata/fedora30/proc/stat (hunk 5 above)
Comment (fearful-symmetry) [reply providers/linux/testdata/fedora30/proc/stat]: Apparently the code will try and find a `/proc/host` file before it does anything else. Without this the test will fail.

File: types/host.go [hunk 10]
```diff
@@ -26,6 +26,33 @@ type Host interface {
...
+	NetworkCounters() (*NetworkCountersInfo, error)
+}
+
+// SNMP represents the data from /proc/net/snmp
+type SNMP struct {
+	IP      map[string]int64 `json:"ip" netstat:"Ip"`
+	ICMP    map[string]int64 `json:"icmp" netstat:"Icmp"`
+	ICMPMsg map[string]int64 `json:"icmp_msg" netstat:"IcmpMsg"`
+	TCP     map[string]int64 `json:"tcp" netstat:"Tcp"`
+	UDP     map[string]int64 `json:"Udp" netstat:"Udp"`
```
Comment (andrewkroh) [inline types/host.go]: I think the JSON tag was meant to be `udp`.

Comment (fearful-symmetry) [reply types/host.go]: Nice catch!

**Pull request merged**